src/core/ext/filters/rbac/rbac_filter.cc

```diff
@@ -41,131 +43,58 @@
-                                 GRPC_STATUS_PERMISSION_DENIED);
+ArenaPromise<ServerMetadataHandle> RbacFilter::MakeCallPromise(
+    CallArgs call_args, NextPromiseFactory next_promise_factory) {
+  // Fetch and apply the rbac policy from the service config.
+  auto* service_config_call_data = static_cast<ServiceConfigCallData*>(
+      GetContext<
+          grpc_call_context_element>()[GRPC_CONTEXT_SERVICE_CONFIG_CALL_DATA]
+          .value);
+  auto* method_params = static_cast<RbacMethodParsedConfig*>(
+      service_config_call_data->GetMethodParsedConfig(
+          service_config_parser_index_));
+  if (method_params == nullptr) {
+    return Immediate(ServerMetadataFromStatus(
+        absl::PermissionDeniedError("No RBAC policy found.")));
+  } else {
+    auto* authorization_engine = method_params->authorization_engine(index_);
+    if (authorization_engine
+            ->Evaluate(EvaluateArgs(call_args.client_initial_metadata.get(),
+                                    &per_channel_evaluate_args_))
+            .type == AuthorizationEngine::Decision::Type::kDeny) {
+      return Immediate(ServerMetadataFromStatus(
+          absl::PermissionDeniedError("Unauthorized RPC rejected")));
     }
   }
-  grpc_closure* closure = calld->original_recv_initial_metadata_ready_;
-  calld->original_recv_initial_metadata_ready_ = nullptr;
-  Closure::Run(DEBUG_LOCATION, closure, error);
+  return next_promise_factory(std::move(call_args));
 }
 
-//
-// RbacFilter
-//
-
-const grpc_channel_filter RbacFilter::kFilterVtable = {
-    RbacFilter::CallData::StartTransportStreamOpBatch,
-    nullptr,
-    grpc_channel_next_op,
-    sizeof(RbacFilter::CallData),
-    RbacFilter::CallData::Init,
-    grpc_call_stack_ignore_set_pollset_or_pollset_set,
-    RbacFilter::CallData::Destroy,
-    sizeof(RbacFilter),
-    RbacFilter::Init,
-    grpc_channel_stack_no_post_init,
-    RbacFilter::Destroy,
-    grpc_channel_next_get_info,
-    "rbac_filter",
-};
+const grpc_channel_filter RbacFilter::kFilterVtable =
+    MakePromiseBasedFilter<RbacFilter, FilterEndpoint::kServer>("rbac_filter");
 
 RbacFilter::RbacFilter(size_t index,
                        EvaluateArgs::PerChannelArgs per_channel_evaluate_args)
     : index_(index),
       service_config_parser_index_(RbacServiceConfigParser::ParserIndex()),
       per_channel_evaluate_args_(std::move(per_channel_evaluate_args)) {}
 
-grpc_error_handle RbacFilter::Init(grpc_channel_element* elem,
-                                   grpc_channel_element_args* args) {
-  GPR_ASSERT(elem->filter == &kFilterVtable);
-  auto* auth_context = args->channel_args.GetObject<grpc_auth_context>();
+absl::StatusOr<RbacFilter> RbacFilter::Create(const ChannelArgs& args,
+                                              ChannelFilter::Args filter_args) {
+  auto* auth_context = args.GetObject<grpc_auth_context>();
   if (auth_context == nullptr) {
     return GRPC_ERROR_CREATE("No auth context found");
```

Comment:
Suggest removing all use of `GRPC_ERROR_CREATE()` in this code at the same time, because nothing in the filter promise API requires use of the old `grpc_error` extensions.

Reply:
I'm going to go ahead and merge and then follow-up later: I'm quickly approaching the point of it being dangerous for me to land PRs prior to vacation.